libkbfs/file_data.go

```diff
@@ -750,54 +1060,120 @@ func (fd *fileData) truncateExtend(ctx context.Context, size uint64,
+//   have been changed yet, but which will need to change as part of the
+//   sync process because of leaf node changes below it.
 // * unrefs: a slice of BlockInfos that must be unreferenced as part of an
 //   eventual sync of this write.  May be non-nil even if err != nil.
 // * newlyDirtiedChildBytes is the total amount of block data dirtied by this
-//   write, including the entire size of blocks that have had at least one
+//   truncate, including the entire size of blocks that have had at least one
 //   byte dirtied.  As above, it may be non-zero even if err != nil.
 func (fd *fileData) truncateShrink(ctx context.Context, size uint64,
 	topBlock *FileBlock, oldDe DirEntry) (
-	newDe DirEntry, unrefs []BlockInfo, newlyDirtiedChildBytes int64,
-	err error) {
+	newDe DirEntry, dirtyPtrs []BlockPointer, unrefs []BlockInfo,
+	newlyDirtiedChildBytes int64, err error) {
 	iSize := int64(size) // TODO: deal with overflow
+
 	ptr, parentBlocks, block, nextBlockOff, startOff, wasDirty, err :=
 		fd.getFileBlockAtOffset(ctx, topBlock, iSize, blockWrite)
 	if err != nil {
-		return DirEntry{}, nil, 0, err
+		return DirEntry{}, nil, nil, 0, err
 	}
 
 	oldLen := len(block.Contents)
-	// We need to delete some data (and possibly entire blocks).
+	// We need to delete some data (and possibly entire blocks).  Note
+	// we make a new slice and copy data in order to make sure the
+	// data being truncated can be fully garbage-collected.
 	block.Contents = append([]byte(nil), block.Contents[:iSize-startOff]...)
 
 	newlyDirtiedChildBytes = int64(len(block.Contents))
 	if wasDirty {
 		newlyDirtiedChildBytes -= int64(oldLen) // negative
 	}
 
+	dirtyMap := make(map[BlockPointer]bool)
 	if nextBlockOff > 0 {
-		// TODO: if indexInParent == 0, we can remove the level of indirection.
-		// TODO: support multiple levels of indirection.
-		parentBlock := parentBlocks[0].pblock
-		indexInParent := parentBlocks[0].childIndex
-		for _, ptr := range parentBlock.IPtrs[indexInParent+1:] {
-			unrefs = append(unrefs, ptr.BlockInfo)
+		// TODO: remove any unnecessary levels of indirection if the
+		// number of leaf nodes shrinks significantly (KBFS-1824).
+
+		// Get all paths to any leaf nodes following the new
+		// right-most block, since those blocks need to be
+		// unreferenced, and their parents need to be modified or
+		// unreferenced.
+		pfr, err := fd.getIndirectBlocksForOffsetRange(
+			ctx, topBlock, nextBlockOff, -1)
+		if err != nil {
+			return DirEntry{}, nil, nil, 0, err
 		}
-		parentBlock.IPtrs = parentBlock.IPtrs[:indexInParent+1]
-		// always make the parent block dirty, so we will sync it
-		if err = fd.cacher(fd.file.tailPointer(), parentBlock); err != nil {
-			return DirEntry{}, nil, newlyDirtiedChildBytes, err
+
+		cachedPtrs := make(map[BlockPointer][]IndirectFilePtr)
+		for _, path := range pfr {
+			// Fake an indirect pointer for the top block.
+			iptr := IndirectFilePtr{BlockInfo: BlockInfo{
```

Comment:
also probably less confusing to name this `parentInfo`, and have a comment saying that it's a loop invariant that `parentInfo.BlockPointer` points to `pb.pblock` in the loop below.